### PR TITLE
1873: production, dividends, buying trains

### PR DESCRIPTION
**Pull request by roseundy**
Common file changes:
* UI::BuyTrain - changes to support scrapping trains, buying trains for subsidiary mines, and paying maintenance upon purchase
* UI::Operating - small change to support scrapping trains via new action
* Action:: BuyTrain - new parameters for "slots" and "extra_due" (first for subsidiary mines, second for when seller pays maintenance costs)
* Action::ScrapTrain - new action to support scrapping trains during the buy_trains step

Changes to 1873 code:
* Various bug fixes
* Calculation of mine revenue (indie and public)
* Calculation of maintenance costs for all
* Dividend step to take maintenance costs and mine revenue into account
* Implement MHE operation
* Major change to buy_train to implement all the various and sundry 1873 train purchase rules, including:
* Creating switchers on the fly
* Creating multiple machines distributing them to subsidiary mines
* Scrapping trains, including auto-scrapping of machines and switchers

1873_new_buy_trains

### CODE REVIEW

File: assets/app/view/game/buy_trains.rb [hunk 1]
```diff
@@ -127,6 +129,20 @@ def render
...
+        @slot_checkboxes = {}
+        if step.respond_to?(:slot_view) && (view = step.slot_view(@corporation))
+          children << send("render_#{view}")
+        end
+
+        scrappable_trains = []
+        scrappable_trains = step.scrappable_trains(@corporation) if step.respond_to?(:scrappable_trains)
+        unless scrappable_trains.empty?
+          children << h(:h3, h3_props, 'Trains to Scrap')
+          children << h(:div, div_props, [
```
Comment (tobymao) [inline assets/app/view/game/buy_trains.rb:141]: scrap_trains(scrappable_trains)

no need to make an array and then explode an array...

File: lib/engine/game/g_1873/game.rb [hunk 2]
```diff
@@ -683,40 +839,148 @@ def pres_change_ok?(corporation)
...
+        end
+
+        # find the best mine slot for a switcher, first based on switcher sizes, then maximizing
+        # revenue
+        def find_switcher_slot(entity, new_size)
+          mines = public_mine_mines(entity)
+          min_size = mines.map { |m| switcher_size(m) || 0 }.min
+          candidates = mines.each_index.map { |i| i }.select { |i| (switcher_size(mines[i]) || 0) == min_size }
+          candidates.max do |c|
+            mines.each_index.map { |i| i }.sum do |i|
```
Comment (tobymao) [inline lib/engine/game/g_1873/game.rb:892]: map seems useless

File: lib/engine/game/g_1873/game.rb [hunk 3]
```diff
@@ -683,40 +839,148 @@ def pres_change_ok?(corporation)
...
+            end
+          end
+        end
+
+        # find the best mine slot for a switcher, first based on switcher sizes, then maximizing
+        # revenue
+        def find_switcher_slot(entity, new_size)
+          mines = public_mine_mines(entity)
+          min_size = mines.map { |m| switcher_size(m) || 0 }.min
+          candidates = mines.each_index.map { |i| i }.select { |i| (switcher_size(mines[i]) || 0) == min_size }
```
Comment (tobymao) [inline lib/engine/game/g_1873/game.rb:890]: why are you doing index  map identity?

File: lib/engine/game/g_1873/game.rb [hunk 4]
```diff
@@ -683,40 +839,148 @@ def pres_change_ok?(corporation)
...
+
+          @corporation_info[entity][:mines]
         end
 
-        def maintenance_costs(_corp)
-          0
+        # find the best mine slot for a machine out of the legal candidates (ones that have a smaller machine)
+        def find_machine_slot(entity, new_size)
+          mines = public_mine_mines(entity)
+          candidates = mines.each_index.map { |i| i }.select { |i| machine_size(mines[i]) < new_size }
```
Comment (tobymao) [inline lib/engine/game/g_1873/game.rb:877]: stop copy pasting this

File: lib/engine/game/g_1873/step/buy_train.rb [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+              t.variants.values.all? { |v| v[:price] > entity.cash }
+            end
+          end
+
+          def buyable_trains(entity)
+            depot_trains = buyable_depot_trains(entity)
+            other_trains = @depot.other_trains(entity)
+            other_trains = [] if entity.cash.zero?
+            other_trains.reject! do |ot|
+              @game.train_is_machine?(ot) || @game.any_mine?(entity) && @game.train_is_train?(ot) ||
```
Comment (tobymao) [inline lib/engine/game/g_1873/step/buy_train.rb:92]: put some parens around this

File: lib/engine/game/g_1873/step/buy_train.rb [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+              @game.switcher(entity)
+            elsif @game.public_mine?(entity)
+              @game.public_mine_mines(entity).any? { |m| @game.switcher(m) }
+            else
+              # RRs can voluntarily scrap switchers
+              # and any trains if they have at least two (the NWE can't scrap a train
+              # if the only one left would be a 1T)
+              return true if entity.trains.any? { |t| @game.train_is_switcher?(t) }
+
+              trains = entity.trains.select { |t| @game.train_is_train?(t) }
```
Comment (tobymao) [inline lib/engine/game/g_1873/step/buy_train.rb:136]: trains.count ... > 1

File: assets/app/view/game/buy_trains.rb (hunk 1 above)
Comment (roseundy) [reply assets/app/view/game/buy_trains.rb:141]: Ok

File: lib/engine/game/g_1873/game.rb (hunk 2 above)
Comment (roseundy) [reply lib/engine/game/g_1873/game.rb:892]: Does sum work on an each?

In any case, I thought I needed these two methods, but I don't.

File: lib/engine/game/g_1873/game.rb (hunk 3 above)
Comment (roseundy) [reply lib/engine/game/g_1873/game.rb:890]: I want an array of indices.

This method is going away.

File: lib/engine/game/g_1873/game.rb (hunk 4 above)
Comment (roseundy) [reply lib/engine/game/g_1873/game.rb:877]: I'm deleting both of these.

File: lib/engine/game/g_1873/step/buy_train.rb (hunk 5 above)
Comment (roseundy) [reply lib/engine/game/g_1873/step/buy_train.rb:92]: Ok

File: lib/engine/game/g_1873/step/buy_train.rb (hunk 6 above)
Comment (roseundy) [reply lib/engine/game/g_1873/step/buy_train.rb:136]: Ok

File: lib/engine/game/g_1873/game.rb (hunk 3 above)
Comment (tobymao) [reply lib/engine/game/g_1873/game.rb:890]: you would do mines.select.with_index instead

Comment (roseundy) [reply lib/engine/game/g_1873/game.rb:890]: Yeah, I missed that you could use with_index on more methods than each. That explains why my code is so ugly.

File: lib/engine/game/g_1873/game.rb (hunk 2 above)
Comment (roseundy) [reply lib/engine/game/g_1873/game.rb:892]: method deleted

File: lib/engine/game/g_1873/game.rb (hunk 3 above)
Comment (roseundy) [reply lib/engine/game/g_1873/game.rb:890]: method deleted

File: lib/engine/game/g_1873/game.rb (hunk 4 above)
Comment (roseundy) [reply lib/engine/game/g_1873/game.rb:877]: method deleted

**Review by tobymao — APPROVED** [review]

**Pull request merged**